runners/gradle-plugin/src/main/kotlin/org/jetbrains/dokka/gradle/kotlin/kotlinClasspathUtils.kt

```diff
@@ -41,12 +48,20 @@ private fun KotlinCompilation.compileClasspathOf(project: Project): FileCollecti
         return this.classpathOf(project)
     }
 
-    val platformDependencyFiles: FileCollection = (this as? AbstractKotlinNativeCompilation)
-        ?.target?.project?.configurations
-        ?.findByName(@Suppress("DEPRECATION") this.defaultSourceSet.implementationMetadataConfigurationName) // KT-58640
-        ?: project.files()
+    val kgpVersion = project.getKgpVersion()
+    val shouldKeepBackwardsCompatibility = (kgpVersion != null && kgpVersion < KotlinGradlePluginVersion(1, 8, 0))
+
+    return if (shouldKeepBackwardsCompatibility) {
+        val platformDependencyFiles: FileCollection = (this as? AbstractKotlinNativeCompilation)
+            ?.target?.project?.configurations
+            ?.findByName(@Suppress("DEPRECATION") this.defaultSourceSet.implementationMetadataConfigurationName) // KT-58640
+            ?: project.files()
 
-    return this.compileDependencyFiles + platformDependencyFiles + this.classpathOf(project)
+        this.compileDependencyFiles + platformDependencyFiles + this.classpathOf(project)
+    } else {
+        // compileTaskProvider is available starting from kotlin 1.8.0
+        (compileTaskProvider.get() as? KotlinCompileTool)?.run { sources + libraries } ?: project.files()
```

Comment:
this is fix for https://github.com/Kotlin/dokka/issues/3068
previous solution uses `compileDependencyFiles` which in itself has dependency on https://github.com/JetBrains/kotlin/blob/b4335c86c8378a43a63052ca00c2472f355d375c/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/sources/KotlinSourceSetMetadataTransformation.kt#L23 (deprecated for removal)

Reply:
I played around with your changes to try and understand why it fixes the problem, as this is also legacy that we carry without fully understanding what it's doing and why.

It looks like the problem is not with `compileDependencyFiles`, but with `platformDependencyFiles`. I see that removing the whole `platformDependencyFiles` block fixes the problem both in your project and in the one from #3068, but I don't quite understand why :smiling_face_with_tear: Maybe you had a chance to research it and have an idea?

I've also noticed that both reproducer projects have quite unusual and non-standard Gradle configuration subprojects, so I wouldn't be surprised if this bug needs specific conditions to be reproduced. I tried to reproduce it on ktor, but they haven't migrated to Kotlin 1.9.0 yet due to some problems. Can't think of any other average multiplatform multi-module project to test this on...

I'll consult with the multiplatform team in the upcoming days, maybe they have an idea about what's going on..

Reply:
> but I don't quite understand why

me neither :)

> I've also noticed that both reproducer projects have quite unusual and non-standard Gradle configuration subprojects

Overall, at least with my project, while it could be unusual, it's mostly recommended by Gradle practice to use small separate convention plugins and apply them to subproject instead of configuration via `allprojects/subprojects` in root repository - this will be the only way to do it if/when project isolation support in Gradle will be implemented fully. Or are you talking about something else unusual?
Just in case, almost full configuration of Dokka for subproject is in one file - https://github.com/whyoleg/cryptography-kotlin/blob/dev/gradle/plugins/build-logic/src/main/kotlin/buildx-documentation.gradle.kts
Which is applied in scope of other convention plugin

Reply:
To expand on the unusual project configuration 

I usually see a flat and simple structure, where you have `buildSrc` / `build-logic` in the root, and then convention plugins inside it. The version catalog is usually a toml file inside `gradle/libs.toml` that gets imported by the simple `buildSrc` / `build-logic`.

cryptography-kotlin and [kmm-vc-library](https://github.com/a-sit-plus/kmm-vc-library/tree/main/conventions-vclib) on the surface level have the same approach, but both seem to have some additional configurations/projects in places you wouldn't expect to find them (like separate projects for the version catalog and the kotlin version)

That's not to say that Dokka is doing everything correctly and there is no bug (there likely is), but knowing how fragile Gradle is and looking at the error message `Cannot change dependencies of dependency configuration .. after task dependencies have been resolved` - I wouldn't be surprised if these specific project configurations are playing a role here

I also don't think there were any other reports about this particular problem, but there for sure must be many multiplatform multi-module projects that use Dokka. So there must be something common between cryptography and kmm-vc-library. Or maybe people haven't updated to 1.9.0 yet and the storm is coming our way (god, I hope not) 😅

Reply:
Yeah, I understand that it's not the most straightforward approach, but still it's rather good supported from gradle side.
Minor off-top why it's done like this:
1. `build-parameters` generates type-safe accessors for custom gradle project properties. Overall it's alternative to similar concept in KGP: [PropertiesProvider](https://github.com/JetBrains/kotlin/blob/6943d03883eed6013d4f8f1557be28a6b324d8dc/libraries/tools/kotlin-gradle-plugin/src/common/kotlin/org/jetbrains/kotlin/gradle/plugin/PropertiesProvider.kt#L18). It's possible to write all those property accessors just with plain string, but it's not a good idea IMO. And it's in separate included build, because it should be used in other included builds :)
2. `kotlin-version-catalog` this is rather specific thing in my projects to simplify testing of new Kotlin releases (like 1.9.0, 1.9.20-dev). It allows to override kotlin version for all places in a project via a single command line argument (`-PuseKotlin=1.9.20`). The idea is similar to what could be archived by using recently created [community-project-gradle-plugin](https://github.com/Kotlin/community-project-gradle-plugin) but using different strait forward approach and integrated inside build, so Im able to easy test new kotlin versions (even dev versions) both locally, in IDEA and on CI. And because it overrides kotlin versions in a lot of places (both runtime dependencies and gradle plugin dependencies), the only way to achieve it is to use included build.

While all of it could be achieved via simpler ways it will increase complexity and possibility to misuse something

P.S. end of off-top :)

Returning to cause of an issue - it's really hard to understand for me what is the real cause on why it worked before and doesn't work now. May be the main difference is that Dokka in my project applied in specific subprojects and not how stated in readme (in root via `subprojects {}`).
Also I see, that some other projects start to adopt https://github.com/adamko-dev/dokkatoo because of better compatibility with new and shiny gradle features :)
just FYI, even `gradle` uses it instead of official Dokka gradle plugin: https://github.com/search?q=repo%3Agradle%2Fgradle%20dokkatoo&type=code